diff -Nru reiserfsprogs-3.6.20.orig/utils/debugfs/Makefile.in reiserfsprogs-3.6.20/utils/debugfs/Makefile.in
--- reiserfsprogs-3.6.20.orig/utils/debugfs/Makefile.in	2008-09-26 23:49:24.000000000 +0200
+++ reiserfsprogs-3.6.20/utils/debugfs/Makefile.in	2008-09-26 23:54:12.000000000 +0200
@@ -590,7 +590,7 @@
 install-debugreiserfs:
 	if test -f $(DESTDIR)$(sbindir)/debugfs.reiserfs; then \
 		rm -f $(DESTDIR)$(sbindir)/debugreiserfs; \
-		ln $(DESTDIR)$(sbindir)/debugfs.reiserfs $(DESTDIR)$(sbindir)/debugreiserfs; \
+		ln -s debugfs.reiserfs $(DESTDIR)$(sbindir)/debugreiserfs; \
 	fi;
 
 install-data-local: install-debugreiserfs
diff -Nru reiserfsprogs-3.6.20.orig/utils/fsck/Makefile.in reiserfsprogs-3.6.20/utils/fsck/Makefile.in
--- reiserfsprogs-3.6.20.orig/utils/fsck/Makefile.in	2008-09-26 23:49:24.000000000 +0200
+++ reiserfsprogs-3.6.20/utils/fsck/Makefile.in	2008-09-26 23:54:22.000000000 +0200
@@ -778,7 +778,7 @@
 install-reiserfsck:
 	if test -f $(DESTDIR)$(sbindir)/fsck.reiserfs; then \
 		rm -f $(DESTDIR)$(sbindir)/reiserfsck; \
-		ln $(DESTDIR)$(sbindir)/fsck.reiserfs $(DESTDIR)$(sbindir)/reiserfsck; \
+		ln -s fsck.reiserfs $(DESTDIR)$(sbindir)/reiserfsck; \
 	fi;
 
 install-data-local: install-reiserfsck
diff -Nru reiserfsprogs-3.6.20.orig/utils/mkfs/Makefile.in reiserfsprogs-3.6.20/utils/mkfs/Makefile.in
--- reiserfsprogs-3.6.20.orig/utils/mkfs/Makefile.in	2008-09-26 23:49:24.000000000 +0200
+++ reiserfsprogs-3.6.20/utils/mkfs/Makefile.in	2008-09-26 23:54:32.000000000 +0200
@@ -491,7 +491,7 @@
 install-mkreiserfs:
 	if test -f $(DESTDIR)$(sbindir)/mkfs.reiserfs; then \
 		rm -f $(DESTDIR)$(sbindir)/mkreiserfs; \
-		ln $(DESTDIR)$(sbindir)/mkfs.reiserfs $(DESTDIR)$(sbindir)/mkreiserfs; \
+		ln -s mkfs.reiserfs $(DESTDIR)$(sbindir)/mkreiserfs; \
 	fi;
 
 install-data-local: install-mkreiserfs
diff -Nru reiserfsprogs-3.6.20.orig/utils/resizer/Makefile.in reiserfsprogs-3.6.20/utils/resizer/Makefile.in
--- reiserfsprogs-3.6.20.orig/utils/resizer/Makefile.in	2008-09-26 23:49:24.000000000 +0200
+++ reiserfsprogs-3.6.20/utils/resizer/Makefile.in	2008-09-26 23:54:42.000000000 +0200
@@ -523,7 +523,7 @@
 install-resize_reiserfs:
 	if test -f $(DESTDIR)$(sbindir)/resizefs.reiserfs; then \
 		rm -f $(DESTDIR)$(sbindir)/resize_reiserfs; \
-		ln $(DESTDIR)$(sbindir)/resizefs.reiserfs $(DESTDIR)$(sbindir)/resize_reiserfs; \
+		ln -s resizefs.reiserfs $(DESTDIR)$(sbindir)/resize_reiserfs; \
 	fi;
 
 install-data-local: install-resize_reiserfs
diff -Nru reiserfsprogs-3.6.20.orig/utils/tune/Makefile.in reiserfsprogs-3.6.20/utils/tune/Makefile.in
--- reiserfsprogs-3.6.20.orig/utils/tune/Makefile.in	2008-09-26 23:49:24.000000000 +0200
+++ reiserfsprogs-3.6.20/utils/tune/Makefile.in	2008-09-26 23:54:50.000000000 +0200
@@ -491,7 +491,7 @@
 install-reiserfstune:
 	if test -f $(DESTDIR)$(sbindir)/tunefs.reiserfs; then \
 		rm -f $(DESTDIR)$(sbindir)/reiserfstune; \
-		ln $(DESTDIR)$(sbindir)/tunefs.reiserfs $(DESTDIR)$(sbindir)/reiserfstune; \
+		ln -s tunefs.reiserfs $(DESTDIR)$(sbindir)/reiserfstune; \
 	fi;
 
 install-data-local: install-reiserfstune
