diff -ru devfsd/devfsd.c devfsd-new/devfsd.c
--- devfsd/devfsd.c	2002-03-25 07:59:40.000000000 +0100
+++ devfsd-new/devfsd.c	2002-05-25 15:35:31.000000000 +0200
@@ -641,10 +641,7 @@
 	fprintf (stderr, "Error setting SIGUSR1 handler\t%s\n", ERRSTRING);
 	exit (1);
     }
-    if (trace_level < 1)
-    	fprintf (stderr, "Started device management daemon v%s for %s\n",
-		 DEVFSD_VERSION, mount_point);
-    else
+    if (trace_level > 0)
     	fprintf (stderr,
 		 "Started device management daemon v%s for %s at trace level %d\n",
 		 DEVFSD_VERSION, mount_point, trace_level);
