--- util-linux-ng-2.14.1.orig/mount/fstab.c	2008-10-01 10:00:15.000000000 +0200
+++ util-linux-ng-2.14.1/mount/fstab.c	2008-10-01 10:03:09.000000000 +0200
@@ -30,9 +30,12 @@
 
 static void
 get_mtab_info(void) {
-	struct stat mtab_stat;
-
 	if (!have_mtab_info) {
+		struct stat mtab_stat;
+
+		var_mtab_does_not_exist = 0;
+		var_mtab_is_a_symlink = 0;
+
 		if (lstat(_PATH_MOUNTED, &mtab_stat))
 			var_mtab_does_not_exist = 1;
 		else if (S_ISLNK(mtab_stat.st_mode))
