diff -ru util-linux-2.12q/MCONFIG util-linux-2.12q-new/MCONFIG
--- util-linux-2.12q/MCONFIG	2004-12-13 00:15:48.000000000 +0100
+++ util-linux-2.12q-new/MCONFIG	2005-02-03 22:28:59.585754816 +0100
@@ -102,7 +102,7 @@
 # If HAVE_SLN is set to "yes", then sln won't be installed
 # (but the man page sln.8 will be installed anyway).
 # sln also comes with libc and glibc.
-HAVE_SLN=no
+HAVE_SLN=yes
 
 # If HAVE_FDUTILS is set to "yes", then setfdprm won't be installed.
 HAVE_FDUTILS=no
@@ -114,7 +114,7 @@
 # If configure decides that Native Language Support (NLS) is available,
 # it sets MAY_ENABLE_NLS in defines.h. If you don't want NLS, set
 # DISABLE_NLS to "yes".
-DISABLE_NLS=no
+DISABLE_NLS=yes
 
 # Different optimizations for different cpus.
 # Before gcc 2.8.0 only -m486
@@ -177,7 +177,7 @@
 BIN_DIR=        /bin
 VAR_PATH=	/var
 LOG_DIR=	/var/log
-MAN_DIR=	/usr/share/man
+MAN_DIR=	/usr/man
 INFO_DIR=	/usr/share/info
 
 #
diff -ru util-linux-2.12q/disk-utils/fsck.cramfs.c util-linux-2.12q-new/disk-utils/fsck.cramfs.c
--- util-linux-2.12q/disk-utils/fsck.cramfs.c	2004-12-11 15:53:16.000000000 +0100
+++ util-linux-2.12q-new/disk-utils/fsck.cramfs.c	2005-02-03 22:28:59.584754968 +0100
@@ -76,7 +76,6 @@
 
 #define PAD_SIZE 512
 
-#include <asm/page.h>
 #ifdef PAGE_SIZE
 #define PAGE_CACHE_SIZE ((int) PAGE_SIZE)
 #elif defined __ia64__
diff -ru util-linux-2.12q/login-utils/agetty.c util-linux-2.12q-new/login-utils/agetty.c
--- util-linux-2.12q/login-utils/agetty.c	2002-07-29 09:36:42.000000000 +0200
+++ util-linux-2.12q-new/login-utils/agetty.c	2005-02-03 22:28:59.586754664 +0100
@@ -72,7 +72,7 @@
 #include <time.h>
 #endif
 
-#define LOGIN " login: "		/* login prompt */
+#define LOGIN "login: "		/* login prompt */
 
 /* Some shorthands for control characters. */
 
@@ -951,14 +951,6 @@
 	(void) fclose(fd);
     }
 #endif
-#ifdef __linux__
-	{
-		char hn[MAXHOSTNAMELEN+1];
-
-		(void) gethostname(hn, MAXHOSTNAMELEN);
-		write(1, hn, strlen(hn));
-	}
-#endif		
     (void) write(1, LOGIN, sizeof(LOGIN) - 1);	/* always show login prompt */
 }
 
