Index: prtget.cpp
===================================================================
--- prtget.cpp	(revision 1875)
+++ prtget.cpp	(working copy)
@@ -931,8 +931,7 @@
         }
     }
 
-    if (m_undefinedVersionComp.size() && 
-        (m_parser->all() || m_parser->verbose()) ) {
+    if (m_undefinedVersionComp.size()) {
         cout << "\n\n" << "Undecidable version differences (use --strict-diff)"
              << endl << endl;
         printFormattedDiffLine("Port",
