--- general.m4.orig	2006-07-07 09:13:11.000000000 -0500
+++ general.m4	2006-07-07 09:13:39.000000000 -0500
@@ -570,7 +570,7 @@
 AC_SUBST([psdir],          ['${docdir}'])dnl
 AC_SUBST([libdir],         ['${exec_prefix}/lib'])dnl
 AC_SUBST([localedir],      ['${datarootdir}/locale'])dnl
-AC_SUBST([mandir],         ['${datarootdir}/man'])dnl
+AC_SUBST([mandir],         ['${prefix}/man'])dnl
 
 ac_prev=
 ac_dashdash=
@@ -1005,7 +1005,7 @@
   --datadir=DIR          read-only architecture-independent data [DATAROOTDIR]
   --infodir=DIR          info documentation [DATAROOTDIR/info]
   --localedir=DIR        locale-dependent data [DATAROOTDIR/locale]
-  --mandir=DIR           man documentation [DATAROOTDIR/man]
+  --mandir=DIR           man documentation [PREFIX/man]
   --docdir=DIR           documentation root ]@<:@DATAROOTDIR/doc/m4_ifset([AC_PACKAGE_TARNAME], [AC_PACKAGE_TARNAME], [PACKAGE])@:>@[
   --htmldir=DIR          html documentation [DOCDIR]
   --dvidir=DIR           dvi documentation [DOCDIR]
