From 4220cc4e0f59c74d9c0c554260523dc9b270f631 Mon Sep 17 00:00:00 2001
From: Tilman Sauerbeck <tilman@crux.nu>
Date: Wed, 13 May 2009 18:45:47 +0200
Subject: [PATCH] Fixed the build with gcc 4.4.

---
 configparser.cpp |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/configparser.cpp b/configparser.cpp
index c70df82..5ed1c95 100644
--- a/configparser.cpp
+++ b/configparser.cpp
@@ -10,6 +10,7 @@
 ////////////////////////////////////////////////////////////////////////
 
 #include <iostream>
+#include <cstdio>
 #include <cstring>
 #include "configparser.h"
 
-- 
1.6.3.1

