diff -Nru bash-3.0/doc/Makefile.in bash-3.0-new/doc/Makefile.in
--- bash-3.0/doc/Makefile.in	2004-07-27 14:57:48.000000000 +0200
+++ bash-3.0-new/doc/Makefile.in	2005-02-02 23:29:13.401509336 +0100
@@ -141,7 +141,7 @@
 	${RM} $@
 	-${DVIPS} $<
 
-all: ps info dvi text html
+all:
 nodvi: ps info text html
 
 PSFILES = bash.ps bashbug.ps article.ps builtins.ps rbash.ps 
@@ -220,21 +220,9 @@
 		$(SHELL) $(SUPPORT_SRCDIR)/mkinstalldirs $(DESTDIR)$(htmldir) ; \
 	fi
 	
-install: info installdirs bash.info
+install: installdirs
 	-$(INSTALL_DATA) $(srcdir)/bash.1 $(DESTDIR)$(man1dir)/bash${man1ext}
 	-$(INSTALL_DATA) $(srcdir)/bashbug.1 $(DESTDIR)$(man1dir)/bashbug${man1ext}
-# uncomment the next line to install the builtins man page
-#	-$(INSTALL_DATA) $(srcdir)/builtins.1 $(DESTDIR)$(man1dir)/bash_builtins${man1ext}
-	-$(INSTALL_DATA) $(srcdir)/bash.info $(DESTDIR)$(infodir)/bash.info
-# run install-info if it is present to update the info directory
-	if $(SHELL) -c 'install-info --version' >/dev/null 2>&1; then \
-		install-info --dir-file=$(DESTDIR)$(infodir)/dir $(DESTDIR)$(infodir)/bash.info; \
-	else true; fi
-# if htmldir is set, install the html files into that directory
-	-if test -n "${htmldir}" ; then \
-		$(INSTALL_DATA) $(srcdir)/bash.html $(DESTDIR)$(htmldir) ; \
-		$(INSTALL_DATA) $(srcdir)/bashref.html $(DESTDIR)$(htmldir) ; \
-	fi
 
 uninstall:
 	-$(RM) $(DESTDIR)$(man1dir)/bash${man1ext} $(DESTDIR)$(man1dir)/bashbug${man1ext}
