diff -ru sysvinit-2.86/src/Makefile sysvinit-2.86-new/src/Makefile
--- sysvinit-2.86/src/Makefile	2004-06-09 14:47:45.000000000 +0200
+++ sysvinit-2.86-new/src/Makefile	2004-12-11 23:19:19.704905960 +0100
@@ -9,7 +9,7 @@
 #
 
 CC	= gcc
-CFLAGS	= -Wall -O2 -fomit-frame-pointer -D_GNU_SOURCE
+CFLAGS	+= -Wall -fomit-frame-pointer -D_GNU_SOURCE
 LDFLAGS	= -s
 STATIC	=
 
@@ -24,11 +24,10 @@
 MAN8	+= shutdown.8 telinit.8
 
 ifeq ($(DISTRO),)
-BIN	+= mountpoint
-SBIN	+= sulogin bootlogd
-USRBIN	+= utmpdump wall
-MAN1	+= mountpoint.1 wall.1
-MAN8	+= sulogin.8 bootlogd.8
+SBIN	+= sulogin
+USRBIN	+= wall
+MAN1	+= wall.1
+MAN8	+= sulogin.8
 endif
 
 ifeq ($(DISTRO),Debian)
@@ -47,7 +46,7 @@
 BIN_GROUP	= root
 BIN_COMBO	= $(BIN_OWNER):$(BIN_GROUP)
 INSTALL		= install -o $(BIN_OWNER) -g $(BIN_GROUP)
-MANDIR		= /usr/share/man
+MANDIR		= /usr/man
 
 # Additional libs for GNU libc.
 ifneq ($(wildcard /usr/lib/libcrypt.a),)
@@ -111,31 +110,32 @@
 
 install:
 		for i in $(BIN); do \
-			$(INSTALL) -m 755 $$i $(ROOT)/bin/; \
+			$(INSTALL) -D -m 755 $$i $(ROOT)/bin/$$i; \
 		done
 		for i in $(SBIN); do \
-			$(INSTALL) -m 755 $$i $(ROOT)/sbin/; \
+			$(INSTALL) -D -m 755 $$i $(ROOT)/sbin/$$i; \
 		done
 		for i in $(USRBIN); do \
-			$(INSTALL) -m 755 $$i $(ROOT)/usr/bin/; \
+			$(INSTALL) -D -m 755 $$i $(ROOT)/usr/bin/$$i; \
 		done
 		# $(INSTALL) -m 755 etc/initscript.sample $(ROOT)/etc/
 		ln -sf halt $(ROOT)/sbin/reboot
 		ln -sf halt $(ROOT)/sbin/poweroff
 		ln -sf init $(ROOT)/sbin/telinit
+		install -d $(ROOT)/bin
 		ln -sf ../sbin/killall5 $(ROOT)/bin/pidof
 		if [ ! -f $(ROOT)/usr/bin/lastb ]; then \
 			ln -sf last $(ROOT)/usr/bin/lastb; \
 		fi
-		$(INSTALL) -m 644 initreq.h $(ROOT)/usr/include/
+		$(INSTALL) -D -m 644 initreq.h $(ROOT)/usr/include/initreq.h
 		for i in $(MAN1); do \
-			$(INSTALL) -m 644 ../man/$$i $(ROOT)$(MANDIR)/man1/; \
+			$(INSTALL) -D -m 644 ../man/$$i $(ROOT)$(MANDIR)/man1/$$i; \
 		done
 		for i in $(MAN5); do \
-			$(INSTALL) -m 644 ../man/$$i $(ROOT)$(MANDIR)/man5/; \
+			$(INSTALL) -D -m 644 ../man/$$i $(ROOT)$(MANDIR)/man5/$$i; \
 		done
 		for i in $(MAN8); do \
-			$(INSTALL) -m 644 ../man/$$i $(ROOT)$(MANDIR)/man8/; \
+			$(INSTALL) -D -m 644 ../man/$$i $(ROOT)$(MANDIR)/man8/$$i; \
 		done
 ifeq ($(ROOT),)
 		#
