Index: libarchive/archive_read_support_compression_xz.c
===================================================================
--- libarchive/archive_read_support_compression_xz.c	(revision 1146)
+++ libarchive/archive_read_support_compression_xz.c	(revision 1147)
@@ -302,11 +302,11 @@
 	 */
 	if (self->code == ARCHIVE_COMPRESSION_XZ)
 		ret = lzma_stream_decoder(&(state->stream),
-		    (1U << 23) + (1U << 21),/* memlimit */
+		    (1U << 30),/* memlimit */
 		    LZMA_CONCATENATED);
 	else
 		ret = lzma_alone_decoder(&(state->stream),
-		    (1U << 23) + (1U << 21));/* memlimit */
+		    (1U << 30));/* memlimit */
 
 	if (ret == LZMA_OK)
 		return (ARCHIVE_OK);
Index: libarchive/archive_read_support_compression_gzip.c
===================================================================
--- libarchive/archive_read_support_compression_gzip.c	(revision 1103)
+++ libarchive/archive_read_support_compression_gzip.c	(revision 1104)
@@ -148,6 +148,7 @@
 		if (p == NULL)
 			return (0);
 		len += ((int)p[len + 1] << 8) | (int)p[len];
+		len += 2;
 	}
 
 	/* Null-terminated optional filename. */

