--- upstream/trunk/src/su.c	2008/09/06 12:51:53	2348
+++ upstream/trunk/src/su.c	2008/11/23 12:10:21	2461
@@ -721,7 +721,8 @@
 	 */
 	if (!amroot && pw_auth (pwent.pw_passwd, name, PW_SU, (char *) 0)) {
 		SYSLOG ((pwent.pw_uid ? LOG_NOTICE : LOG_WARN,
-			 "Authentication failed for %s", name));
+		         "Authentication failed for %s", name));
+		fprintf(stderr, _("%s: Authentication failure\n"), Prog);
 		su_failure (tty);
 	}
 	(void) signal (SIGQUIT, oldsig);
