diff -Nru shadow-4.1.0.orig/lib/encrypt.c shadow-4.1.0/lib/encrypt.c
--- shadow-4.1.0.orig/lib/encrypt.c	2008-01-22 18:32:09.000000000 +0100
+++ shadow-4.1.0/lib/encrypt.c	2008-01-22 18:35:08.000000000 +0100
@@ -32,6 +32,7 @@
 #ident "$Id: encrypt.c 1457 2007-11-24 00:37:37Z nekral-guest $"
 
 #include <unistd.h>
+#include <stdio.h>
 #include "prototypes.h"
 #include "defines.h"
 char *pw_encrypt (const char *clear, const char *salt)
diff -Nru shadow-4.1.0.orig/libmisc/salt.c shadow-4.1.0/libmisc/salt.c
--- shadow-4.1.0.orig/libmisc/salt.c	2008-01-22 18:32:09.000000000 +0100
+++ shadow-4.1.0/libmisc/salt.c	2008-01-22 18:32:50.000000000 +0100
@@ -13,6 +13,7 @@
 
 #include <sys/time.h>
 #include <stdlib.h>
+#include <stdio.h>
 #include <assert.h>
 #include "prototypes.h"
 #include "defines.h"
diff -Nru shadow-4.1.0.orig/src/usermod.c shadow-4.1.0/src/usermod.c
--- shadow-4.1.0.orig/src/usermod.c	2008-01-22 18:32:09.000000000 +0100
+++ shadow-4.1.0/src/usermod.c	2008-01-22 18:37:06.000000000 +0100
@@ -1146,7 +1146,9 @@
 	pw_locked = 0;
 	spw_locked = 0;
 	gr_locked = 0;
+#ifdef SHADOWGRP
 	sgr_locked = 0;
+#endif
 
 	/*
 	 * Close the DBM and/or flat files
