diff -Nru shadow-4.0.5/libmisc/pwdcheck.c shadow-4.0.5-new/libmisc/pwdcheck.c
--- shadow-4.0.5/libmisc/pwdcheck.c	2004-06-03 01:50:10.000000000 +0200
+++ shadow-4.0.5-new/libmisc/pwdcheck.c	2004-10-29 21:28:31.545167952 +0200
@@ -12,6 +12,7 @@
 #ifdef USE_PAM
 #include "pam_defs.h"
 #endif
+#include <stdio.h>
 #define WRONGPWD2	"incorrect password for `%s'"
 void
 passwd_check (const char *user, const char *passwd, const char *progname)
diff -Nru shadow-4.0.5/src/login.c shadow-4.0.5-new/src/login.c
--- shadow-4.0.5/src/login.c	2004-08-04 12:33:43.000000000 +0200
+++ shadow-4.0.5-new/src/login.c	2004-10-29 21:29:20.704694568 +0200
@@ -812,13 +812,8 @@
 			}
 			preauth_flag = 0;
 #ifndef LOGIN_PROMPT
-#ifdef __linux__		/* hostname login: - like in util-linux login */
-			login_prompt (_("\n%s login: "), username,
-				      sizeof username);
-#else
 			login_prompt (_("login: "), username,
 				      sizeof username);
-#endif
 #else
 			login_prompt (LOGIN_PROMPT, username,
 				      sizeof username);
