--- src/useradd.c.orig	2006-03-29 20:45:56.981071520 +0200
+++ src/useradd.c	2006-03-29 22:39:13.312872040 +0200
@@ -1765,12 +1765,14 @@
 		find_new_uid ();
 	}
 
+#if 0 /* jw */
 	/* do we have to add a group for that user? This is why we need to
 	 * open the group files in the open_files() function  --gafton */
 	if (!(nflg || gflg)) {
 		find_new_gid ();
 		grp_add ();
 	}
+#endif
 
 	usr_update ();
 
