--- Makefile.orig	2007-02-22 09:10:41.000000000 -0600
+++ Makefile	2007-02-22 09:13:35.000000000 -0600
@@ -1,5 +1,4 @@
 # Path to parent kernel include files directory
-KERNEL_INCLUDE=/usr/src/linux/include
 LIBC_INCLUDE=/usr/include
 
 DEFINES= 
@@ -17,12 +16,6 @@
   GLIBCFIX=-Iinclude-glibc -include include-glibc/glibc-bugs.h
 endif
 
-ifeq ($(KERNEL_INCLUDE)/linux/pfkeyv2.h,$(wildcard $(KERNEL_INCLUDE)/linux/pfkeyv2.h))
-  SUBDIRS=libipsec setkey
-  LDLIBS+=-Llibipsec -lipsec
-  IPSECDEF=-DDO_IPSEC -Ilibipsec
-endif
-
 
 #options if you compile with libc5, and without a bind>=4.9.4 libresolv
 # NOT AVAILABLE. Please, use libresolv.
@@ -37,7 +30,7 @@
 IPV6_TARGETS=tracepath6 traceroute6 ping6
 TARGETS=$(IPV4_TARGETS) $(IPV6_TARGETS)
 
-all: check-kernel subdirs $(TARGETS)
+all: subdirs $(TARGETS)
 
 
 tftpd: tftpd.o tftpsubs.o
